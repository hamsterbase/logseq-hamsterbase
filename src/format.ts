import { stringToUUIDFormat } from "./utils";

export type Block = {
  content: string;
  children?: Block[];
  properties?: Record<string, string>;
};

export type WebpageWithHighlight = {
  title: string;
  link?: string;
  id: string;
  highlights: {
    highlightId: string;
    text: string;
    note?: string;
  }[];
  highlight?: {
    highlightId: string;
    text: string;
    note?: string;
  };
};

function replaceContent(webpage: WebpageWithHighlight, content: string) {
  return content.replace(/`{{(.*?)}}`/g, (replace, key) => {
    const keys = key.split(".");
    let value: any = webpage;
    for (const key of keys) {
      value = value?.[key];
    }
    if (!value) {
      return "";
    }
    return String(value);
  });
}

export function formatBlock(
  webpage: WebpageWithHighlight,
  template: Block[]
): Block[] {
  const blocks = template.map((o): Block | Array<Block> => {
    if (o.content === "`{{highlights}}`") {
      return webpage.highlights.map((highlight): Block => {
        return {
          content: replaceContent(
            { ...webpage, highlight },
            o.children![0].content
          ),
          properties: {
            id: stringToUUIDFormat(highlight.highlightId),
          },
        };
      });
      //
    }
    return {
      content: replaceContent(webpage, o.content),
      children: o.children ? formatBlock(webpage, o.children) : [],
    };
  });
  let result: Block[] = [];
  for (const iterator of blocks) {
    if (Array.isArray(iterator)) {
      result = result.concat(iterator);
    } else {
      result.push(iterator);
    }
  }
  return result;
}

export const defaultTemplate = [
  {
    content: "## metadata",
    children: [
      {
        content: "title: `{{title}}`",
      },
      {
        content: "link: `{{link}}`",
      },
    ],
  },
  {
    content: "## highlights",
    children: [
      {
        content: "`{{highlights}}`",
        children: [
          {
            content: "> `{{highlight.text}}` \n\n `{{highlight.note}}`",
          },
        ],
      },
    ],
  },
];
