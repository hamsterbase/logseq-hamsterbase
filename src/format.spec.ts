import { expect, it } from "vitest";
import { formatBlock } from "./format";

it("", () => {
  const res = formatBlock(
    {
      id: "9c75f658c89b0c8580375f51ca843ccdcad3c8b402cc99b9cacdd4c4f8f665f6",
      title: "Don’t think to write, write to think - Herbert Lui",
      link: "https://herbertlui.net/dont-think-to-write-write-to-think/",
      highlights: [
        {
          highlightId: "0000000010_d3692df7-fd33-424a-8a1b-aaa5e9a7cfde",
          text: "Writing is not the artifact of thinking, it’s the actual thinking process.",
        },
      ],
    },
    [
      {
        content: "## metadata",
        children: [
          {
            content: "title: `{{title}}`",
          },
          {
            content: "link: `{{link}}`",
          },
        ],
      },
      {
        content: "## highlights",
        children: [
          {
            content: "`{{highlights}}`",
            children: [
              {
                content: "> `{{highlight.text}}` \n `{{highlight.note}}`",
              },
            ],
          },
        ],
      },
    ]
  );

  expect(JSON.stringify(res, null, 2)).toMatchSnapshot("default-template");
});
