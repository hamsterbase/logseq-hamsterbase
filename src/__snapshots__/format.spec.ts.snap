// Vitest Snapshot v1

exports[` > default-template 1`] = `
"[
  {
    \\"content\\": \\"## metadata\\",
    \\"children\\": [
      {
        \\"content\\": \\"title: Don’t think to write, write to think - Herbert Lui\\",
        \\"children\\": []
      },
      {
        \\"content\\": \\"link: https://herbertlui.net/dont-think-to-write-write-to-think/\\",
        \\"children\\": []
      }
    ]
  },
  {
    \\"content\\": \\"## highlights\\",
    \\"children\\": [
      {
        \\"content\\": \\"> Writing is not the artifact of thinking, it’s the actual thinking process. \\\\n \\",
        \\"properties\\": {
          \\"id\\": \\"3ae86c9f-61c6-50f9-a0c0-ef770ac0b4e6\\"
        }
      }
    ]
  }
]"
`;
